### DISCUSSION TITLE
What kind of tasks to use Asyncify for?

### DISCUSSION THREAD

**Discussion by MarcSkovMadsen** [fastapi/asyncer#107] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Asyncer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Asyncer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Asyncer but to [AnyIO](https://github.com/agronholm/anyio).
- [X] I already checked if it is not related to Asyncer but to [Trio](https://github.com/python-trio/trio).

### Description

I would like to describe `Asyncer` and `asyncify` in the `async` documentation of [HoloViz Panel](https://panel.holoviz.org/reference/index.html) as a very useful tool.

If possible I would like to guide my users as to when `asyncify` is a great solution. I had hoped to find this information in the `Asyncer` docs. But was not able to.

**What kind of tasks does `àsyncify` work best with? And why? How can I best guide the users of Panel?**

## Additional Context

Our users typical use case is that they 

- load some data from a database, a file share, a blob storage or rest api
- process the data using numpy, pandas, polars, scipy, scikit-learn etc.
- create one or more visualizations from the processed data

Panel runs on top of the Bokeh/ Tornado server and really shines when you start using async. But most of the code our users use is sync code. So being able to easily turn that into async code is very valuable.

Normally its recommended to use `async` for io bound tasks and not cpu bound tasks. But in my experience it can be beneficial to use it with Panel for both io and cpu bound tasks. And that is also what I would recommend our users until I know better :-)

Finally I would recommend my users to send tasks to Dask asynchronously if they start experiencing that they have too many cpu bound, async tasks running in the Panel server.

**Comment by j-carson** [comment]
Hi @MarcSkovMadsen 

I am using asyncer with panel trying to follow the docs you wrote, assuming it's  https://panel.holoviz.org/how_to/concurrency/sync_to_async.html

I am adding async functions to my panel app for the first time.  I have existing data-fetch-and-format functions which call an API using the requests library. These functions are not async and are decorated with @pn.cache

Here is what my code looks like:

```python
class MyClass:
     def __init__(self):
        ... # stuff
        self.lock = asyncio.Lock()

    async def data_fetch_method(self, args):
        async with self.lock:
            table1_task = asyncio.create_task( asyncer.asyncify( get_from_table_1 )( args))
            table2_task = asyncio.create_task( asyncer.asyncify( get_from_table_2)(args))

            self.table1_data, self.table2_data = asyncio.gather( table1_task, table2_task) 
```

The questions I don't think the docs answer/demonstrate:

1. When I let my user interface continue to be responsive, the dataset selections can change, so I locked the fetch with an asyncio.Lock(), but it looks like the job is moved to a "worker thread" -- should I use thread-level locking?
2. Do you only use this from within an async callback? I had a lot of problems with trying to make data_fetch_method non-async, so I followed the docs example and used asyncify within an async method instead.  If I made the data-fetch function itself async, can I still use the @pn.cache decorator on it? 
3. Now that the widget in the example can change while the "long task" is running, how do you make sure to not run the task for a value in the widget that is maybe stale? Once I finish the long data fetch, I want to plot it using the most recent user-interface selections, which may be changing now that the app remains responsive.
